File: PiDayAndroid/app/src/main/java/academy/glasscode/piday/core/domain/CalendarFeaturedNumber.kt
package academy.glasscode.piday.core.domain

import java.time.LocalDate

/**
 * The active "digit universe" for the app (π / τ / e / φ / Planck).
 *
 * Requirement: all numbers are treated as equals.
 * The entire app (calendar, detail lookup, stats, battle, share) runs against the
 * selected featured number's bundled index.
 */
enum class CalendarFeaturedNumber(
    val rawValue: String,
    val logoSymbol: String,
    val heatMapSymbol: String,
    val title: String,
    val decimalPreview: String,
    val highlightMonth: Int,
    val highlightDay: Int,
) {
    PI(
        rawValue = "pi",
        logoSymbol = "π",
        heatMapSymbol = "π",
        title = "Pi",
        decimalPreview = "3.14159…",
        highlightMonth = 3,
        highlightDay = 14
    ),
    PI416(
        rawValue = "pi416",
        logoSymbol = "π",
        heatMapSymbol = "π",
        title = "Pi (4.16)",
        decimalPreview = "3.1416…",
        highlightMonth = 4,
        highlightDay = 16
    ),
    TAU(
        rawValue = "tau",
        logoSymbol = "τ",
        heatMapSymbol = "τ",
        title = "Tau",
        decimalPreview = "6.28318…",
        highlightMonth = 6,
        highlightDay = 28
    ),
    EULER(
        rawValue = "e",
        logoSymbol = "e",
        heatMapSymbol = "e",
        title = "Euler",
        decimalPreview = "2.71828…",
        highlightMonth = 2,
        highlightDay = 7
    ),
    GOLDEN_RATIO(
        rawValue = "phi",
        logoSymbol = "φ",
        heatMapSymbol = "φ",
        title = "Phi",
        decimalPreview = "1.61803…",
        highlightMonth = 11,
        highlightDay = 9
    ),
    PLANCK(
        rawValue = "planck",
        logoSymbol = "h",
        heatMapSymbol = "h",
        title = "Planck",
        decimalPreview = "6.62607×10⁻³⁴…",
        highlightMonth = 4,
        highlightDay = 14
    );

    val indexResourceName: String
        get() = when (this) {
            PI, PI416 -> "pi_2026_2035_index"
            TAU -> "tau_2026_2035_index"
            EULER -> "e_2026_2035_index"
            GOLDEN_RATIO -> "phi_2026_2035_index"
            PLANCK -> "planck_2026_2035_index"
        }

    val observedDayLabel: String
        get() = String.format("%02d/%02d", highlightMonth, highlightDay)

    fun highlights(date: LocalDate): Boolean =
        date.monthValue == highlightMonth && date.dayOfMonth == highlightDay

    companion object {
        fun fromRaw(raw: String?): CalendarFeaturedNumber? =
            entries.firstOrNull { it.rawValue == raw }
    }
}

File: PiDay/Core/Domain/CalendarFeaturedNumber.swift
import Foundation

// WHY: The calendar can render multiple digit-heat-map indexes (π, τ, e, φ, Planck).
// Keeping the metadata here gives the ViewModel and Views one shared source of truth
// for titles, date mapping, resource names, and explanatory copy.
enum CalendarFeaturedNumber: String, CaseIterable, Identifiable {
    case pi
    case pi416
    case tau
    case euler
    case goldenRatio
    case planck

    var id: String { rawValue }

    var title: String {
        switch self {
        case .pi: return "Pi"
        case .pi416: return "Pi (4.16)"
        case .tau: return "Tau"
        case .euler: return "e"
        case .goldenRatio: return "Phi"
        case .planck: return "Planck"
        }
    }

    var shortLabel: String {
        switch self {
        case .pi: return "3.14"
        case .pi416: return "4.16"
        case .tau: return "6.28"
        case .euler: return "2.71"
        case .goldenRatio: return "1.61"
        case .planck: return "4.14"
        }
    }

    var logoSymbol: String {
        switch self {
        case .pi, .pi416:
            return "π"
        case .tau:
            return "τ"
        case .euler:
            return "e"
        case .goldenRatio:
            return "φ"
        case .planck:
            return "h"
        }
    }

    var decimalPreview: String {
        switch self {
        case .pi:
            return "3.14159..."
        case .pi416:
            return "4.16"
        case .tau:
            return "6.28318..."
        case .euler:
            return "2.71828..."
        case .goldenRatio:
            return "1.61803..."
        case .planck:
            return "4.13567... × 10⁻¹⁵"
        }
    }

    var heroCopy: String {
        switch self {
        case .pi:
            return "Choose a day to see how early it shows up in the bundled digits of π."
        case .pi416:
            return "Choose a day to see how early it shows up in the bundled digits of π. (Pi Day, but for the 3.1416 crowd.)"
        case .tau:
            return "Choose a day to see how early it shows up in the bundled digits of τ. Tau Day is June 28 (6.28)."
        case .euler:
            return "Choose a day to see how early it shows up in the bundled digits of e. (A nod to 2.71828...)"
        case .goldenRatio:
            return "Choose a day to see how early it shows up in the bundled digits of φ. (A nod to 1.61803...)"
        case .planck:
            return "Choose a day to see how early it shows up in the bundled digits of the Planck constant (h, in eV·s)."
        }
    }

    var badgeTitle: String {
        "Heat Map"
    }

    var badgeValue: String {
        "Exact \(heatMapSymbol) hits"
    }

    var observedDayLabel: String {
        switch self {
        case .pi: return "March 14"
        case .pi416: return "April 16"
        case .tau: return "June 28"
        case .euler: return "February 7"
        case .goldenRatio: return "January 6"
        case .planck: return "April 14"
        }
    }

    var legendText: String {
        switch self {
        case .pi, .pi416:
            return "Based on earliest position in the bundled π digit index."
        case .tau:
            return "Based on earliest position in the bundled τ digit index."
        case .euler:
            return "Based on earliest position in the bundled e digit index."
        case .goldenRatio:
            return "Based on earliest position in the bundled φ digit index."
        case .planck:
            return "Based on earliest position in the bundled Planck-constant digit index."
        }
    }

    var accessibilityHighlightLabel: String {
        switch self {
        case .pi:
            return "classic Pi Day"
        case .pi416:
            return "alternate Pi Day"
        case .tau:
            return "Tau Day"
        case .euler:
            return "Euler's number day"
        case .goldenRatio:
            return "golden ratio day"
        case .planck:
            return "World Quantum Day"
        }
    }

    var usesHeatMap: Bool {
        // NOTE: Whether the heat map is *visible* is determined by whether the
        // corresponding index data is bundled + loaded at runtime.
        true
    }

    // The bundled JSON index file name (without extension) for this constant.
    //
    // pi416 intentionally reuses the pi index — only the celebration framing differs.
    var indexResourceName: String {
        switch self {
        case .pi, .pi416:
            return "pi_2026_2035_index"
        case .tau:
            return "tau_2026_2035_index"
        case .euler:
            return "e_2026_2035_index"
        case .goldenRatio:
            return "phi_2026_2035_index"
        case .planck:
            return "planck_2026_2035_index"
        }
    }

    // Short label for UI copy like "Earlier in …".
    var heatMapSymbol: String {
        switch self {
        case .pi, .pi416:
            return "π"
        case .tau:
            return "τ"
        case .euler:
            return "e"
        case .goldenRatio:
            return "φ"
        case .planck:
            return "h"
        }
    }

    func highlightDate(inYear year: Int, calendar: Calendar) -> Date? {
        let components: DateComponents
        switch self {
        case .pi:
            components = DateComponents(year: year, month: 3, day: 14)
        case .pi416:
            components = DateComponents(year: year, month: 4, day: 16)
        case .tau:
            components = DateComponents(year: year, month: 6, day: 28)
        case .euler:
            components = DateComponents(year: year, month: 2, day: 7)
        case .goldenRatio:
            components = DateComponents(year: year, month: 1, day: 6)
        case .planck:
            components = DateComponents(year: year, month: 4, day: 14)
        }
        return calendar.date(from: components)
    }

    var highlightMonth: Int {
        switch self {
        case .pi: return 3
        case .pi416: return 4
        case .tau: return 6
        case .euler: return 2
        case .goldenRatio: return 1
        case .planck: return 4
        }
    }

    var highlightDay: Int {
        switch self {
        case .pi: return 14
        case .pi416: return 16
        case .tau: return 28
        case .euler: return 7
        case .goldenRatio: return 6
        case .planck: return 14
        }
    }

    func highlights(_ date: Date, calendar: Calendar) -> Bool {
        let year = calendar.component(.year, from: date)
        guard let highlightDate = highlightDate(inYear: year, calendar: calendar) else { return false }
        return calendar.isDate(date, inSameDayAs: highlightDate)
    }
}
